docs/design-proposals/global-scheduler/GlobalSchedulerArchitectureDesignSpecificationFirstDraft.md

```diff
@@ -0,0 +1,416 @@
+
+Our optimistic approach increases parallelism and can support high scalability and low latency performance because all the Global Schedulers operate completely in parallel and do not have to wait for each other, and there is no inter-scheduler head of line blocking.
+
+### 2.5.2 Multi-Dimensional Optimization Based Scheduling Algorithm
+
+The scheduling algorithm consists of two phases. The first phase is filtering. The second phase is scoring.
+
+Different VMs/Containers have different resource requirements and constraints. Factors that need to be considered for filtering include the following:
+
+1. resource requirements, such as number of vCPU, memory size, number of EIPs, disk volume
+2. hardware/software/policy constraints, such as type of disk, GPU
+3. affinity and anti-affinity specifications, such as collocated in the same AZ as another VM/Container/Storage/DB or in a separate AZ from another VM for fault-tolerant purpose
+4. data locality, such as specific geolocation or a specific AZ.
+
+All the clusters need to be filtered according to the VM/Container's specific requirements. This filtering phase will select a list of candidate clusters that are suitable to host/run the VM/Container. If none of the clusters are suitable, the VM/Container remains unscheduled until a Global Scheduler can place it.
+
+Validating a VM/Container's specification of hardware/software/policy constraints, affinity, anti-affinity, and geolocation/AZ locality against a cluster is straightforward. But checking whether a cluster can meet the VM/Container's resource requirements of vCPU, memory, EIPs, and disk volume is tricky. Let's call the latter type of check "_quantity check_".
+
+Our global scheduler does the scheduling/placement of a VM/Container to the granularity of a cluster. To avoid a huge DB, we should not store every node's allocable capacity information in the global DB. But if we only store each cluster's accumulated allocable capacity, "_quantity check_" cannot be done properly. This is because even a cluster's accumulated allocable capacity is larger than a VM/Container's quantity requirement, the cluster may have no single node which has enough resource to host/run the VM/Container. An "average per node allocable capacity" is a better way. But "_quantity check_" against the "average per node allocable capacity" is not good enough either. As shown in Figure 2, the new VM requires 128M memory and Cluster1's "average per node available memory resource" is 110M memory. Although Node 3 in Cluster 1 has 256M, which meets the VM's memory requirement, the "_quantity check_" against Cluster 1 will fail and Cluster 1 will be incorrectly filtered out of the candidate list.
```

Comment:
1、That means global scheduler must collect quantity information of all nodes, and  that conflict with the aim  which we don’t want to have a  huge DB.
2、	What the finish condition of  scan the  largest available node?  Is ’t just a quantity filter?
3、	If  we  cannot get the detail like NUMA info，we   may not avoid the case that the  total quantity enough but  actually cannot host the VM .

Reply:
No, it does not need to collect info of all nodes, only info of the largest node which the cluster provides.

Reply:
How to defined the largest? I mean what is the terminate condition when i rank the host.

Reply:
As stated in the document, we will collect the resource information of several largest available nodes from each cluster (a few with the largest CPU, a few with the largest memory, etc.). Let's call these nodes "potentially feasible nodes". Then the filtering algorithm will run through these "potentially feasible nodes" to select candidate list.

Reply:
We need to deployment a group of resources on the edge  like   2VM of  4U8G ,do we only use 4U8G to match the largest host or use the total cpus 8U16G?

Reply:
Are you talking about bacth scheduling? If these two VMs are assigned to the same Home Scheduler partition, then the 8U16G will be used. If there is no cluster whose lagest node has this capacity, then further check will be done to see if there are two clusters in this partition whose largest nodes both have 4U8G capacity.

Reply:
I suggest to use local-agent do the capacity estimate, it can use the detail infomation of each node  or  each NUMA  to improve the accurancy ,and do not need to put the stats to global scheduler directly .

Reply:
Yes, I am thinking about using local agent too.